src/javascript/expression.rs

```diff
@@ -210,6 +243,215 @@ impl<'module> Generator<'module> {
+                    concat(path.into_iter().map(|i| i.clone().surround("[", "]"))),
+                    ";",
+                    line()
+                ]);
+                Ok(())
+            }
+            Pattern::Discard { .. } => Ok(()),
+            Pattern::Assign { name, pattern, .. } => {
+                assignments.push(docvec![
+                    "let ",
+                    self.next_local_var_name(name),
+                    " = gleam$tmp",
+                    concat(
+                        path.clone()
+                            .into_iter()
+                            .map(|i| i.clone().surround("[", "]"))
+                    ),
+                    ";",
+                    line()
+                ]);
+                self.traverse_pattern(pattern, path, checks, assignments)
+            }
+
+            Pattern::List { elements, tail, .. } => {
+                let path_string = concat(path.into_iter().map(|i| i.clone().surround("[", "]")));
+                let length_check = match tail {
+                    Some(_) => "?.length !== undefined".to_doc(),
+                    None => "?.length === 0".to_doc(),
+                };
+                checks.push(docvec![
+                    "gleam$tmp",
+                    path_string,
+                    Document::String("?.[1]".repeat(elements.len())),
+                    length_check,
+                ]);
+                let _ = elements
+                    .into_iter()
+                    .enumerate()
+                    .map(|x| {
+                        let (index, pattern) = x;
+                        let mut path = path.clone();
+                        for _ in 0..index {
+                            path.push("1".to_doc());
+                        }
+                        path.push("0".to_doc());
+
+                        self.traverse_pattern(pattern, &mut path, checks, assignments)
+                    })
+                    .collect::<Result<Vec<()>, Error>>()?;
+                if let Some(pattern) = tail {
+                    let mut path = path.clone();
+                    for _ in 0..elements.len() {
+                        path.push("1".to_doc());
+                    }
+                    self.traverse_pattern(pattern, &mut path, checks, assignments)?
+                }
+                Ok(())
+            }
+            Pattern::Tuple { elems, .. } => {
+                // We don't check the length, because type system means it's a tuple
+                let _ = elems
+                    .into_iter()
+                    .enumerate()
+                    .map(|x| {
+                        let (index, pattern) = x;
+                        let mut path = path.clone();
+                        path.push(Document::String(format!("{}", index)));
+                        self.traverse_pattern(pattern, &mut path, checks, assignments)
+                    })
+                    .collect::<Result<Vec<()>, Error>>()?;
```

Comment:
Are we building vectors and discarding them? We could use a loop and avoid the work of building the vectors in this function